core/src/main/scala/akka/persistence/cassandra/cleanup/Cleanup.scala

```diff
@@ -42,18 +47,169 @@ final class Cleanup(systemProvider: ClassicActorSystemProvider, settings: Cleanu
   private implicit val askTimeout: Timeout = operationTimeout
 
+  private lazy val session: CassandraSession =
+    CassandraSessionRegistry(system).sessionFor(pluginLocation)
+
+  private lazy val pluginSettings = PluginSettings(system, system.settings.config.getConfig(pluginLocation))
+  private lazy val statements = new CassandraSnapshotStatements(pluginSettings.snapshotSettings)
+  private lazy val selectLatestSnapshotsPs = session.prepare(statements.selectLatestSnapshotMeta)
+  private lazy val selectAllSnapshotMetaPs = session.prepare(statements.selectAllSnapshotMeta)
+
+  private def issueSnapshotDelete(
+      persistenceId: String,
+      maxToKeep: Long,
+      rows: Seq[Row]): Future[Option[SnapshotMetadata]] = {
+    rows match {
+      case Nil =>
+        log.debug("persistence id [{}] has 0 snapshots, no deletes issued", persistenceId)
+        Future.successful(None)
+      case fewer if fewer.size < maxToKeep =>
+        // no delete required, return the oldest snapshot
+        log.debug("Fewer than snapshots than requested for persistence id [{}], no deletes issued", persistenceId)
+        Future.successful(
+          Some(SnapshotMetadata(persistenceId, fewer.last.getLong("sequence_nr"), fewer.last.getLong("timestamp"))))
+      case more =>
+        if (log.isDebugEnabled) {
+          log.debug(
+            "Latest {} snapshots for persistence id [{}] range from {} to {}",
+            maxToKeep,
+            persistenceId,
+            more.head.getLong("sequence_nr"),
+            more.last.getLong("sequence_nr"))
+        }
+        session
+          .executeWrite(statements.deleteSnapshotsBefore, persistenceId, more.last.getLong("sequence_nr"): JLong)
+          .map(_ =>
+            Some(SnapshotMetadata(persistenceId, more.last.getLong("sequence_nr"), more.last.getLong("timestamp"))))
+
+    }
+  }
+
+  /**
+   * Keep all snapshots that occurred after `keepAfter`.
+   * If fewer than `snapshotsToKeep` occurred after `keepAfter` at least that many
+   * are kept. Setting this to 1 ensures that at least snapshot is kept even if it
+   * is older than the `keepAfter`
+   *
+   * If only N number of snapshot should be kept prefer overload without timestamp
+   * as it is more efficient.
+   *
+   * The returned snapshot metadata can be used to issue deletes for events older than the oldest
+   * snapshot.
+   *
+   * @return the snapshot meta of the oldest remaining snapshot. None if there are no snapshots
+   *
+   */
+  def deleteBeforeSnapshot(
+      persistenceId: String,
+      snapshotsToKeep: Int,
+      keepAfter: Long): Future[Option[SnapshotMetadata]] = {
+    require(snapshotsToKeep >= 1, "must keep at least one snapshot")
+    require(keepAfter >= 0, "keepAfter must be greater than 0")
+    selectAllSnapshotMetaPs
+      .flatMap { ps =>
+        val allRows: Source[Row, NotUsed] = session.select(ps.bind(persistenceId))
+        @volatile var count = 0
+        allRows
+          .takeWhile { row =>
```

Comment:
could we skip the count hack with:
```
allRows
  .take(snapshotsToKeep)
  .takeWhile { row =>
```

?

Reply:
No as we may wan to keep more than snapshotsToKeep if they are all newer than the toKeep time